### PR TITLE
[dbnode] Optimize `filesetFiles` function

### PR DESCRIPTION
**Pull request by soundvibe**
**What this PR does / why we need it**:

For namespaces with long retentions huge CPU spikes could emerge during node bootstrap. Looking at the CPU profile, we can see that `filesetFiles` function is not very efficient. 
image
This PR optimizes `filesetFiles` function by reducing the amount of work it needs to do. It now collects `blockStart` and `volumeIndex` fields and later reuse them during sorting and further iterations (without a need to parse them again).

I've written a small benchmark to measure the new implementation (new implementation is ~ 30% faster):
 * 230140 ns/op  - new implementation
 * 324493 ns/op  - old implementation

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note

```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note

```

### CODE REVIEW

File: src/dbnode/persist/fs/files.go
Comment (vpranckaitis) [inline src/dbnode/persist/fs/files.go:1290]: `dataFileSetFilesByTimeAndVolumeIndexAscending` and `fileSetFilesByTimeAndVolumeIndexAscending` seems to be not used anymore after these changes, should be able to remove them, too.

File: src/dbnode/persist/fs/files.go
Comment (vpranckaitis) [inline src/dbnode/persist/fs/files.go:1285]: nit: 

```suggestion
	result := make([]filesetFile, 0, len(matched))
	for _, file := range matched {
		blockStart, volume, err := fn(file)
		if err != nil {
			return nil, err
		}

		result = append(result, filesetFile{
			fileName:    file,
			blockStart:  blockStart,
			volumeIndex: volume,
		})
	}
```

**Review by vpranckaitis — APPROVED** [review]
LGTM 👍 Though would be good to get another review by someone else before merging

**Review by linasm — COMMENTED** [review]
Would you care to share the benchmark code (would be enough to put it in PR description)? 
I am surprised that the improvement is only 30%. What was the sample size for benchmarking?

**Comment by soundvibe** [comment]
> Would you care to share the benchmark code (would be enough to put it in PR description)? I am surprised that the improvement is only 30%. What was the sample size for benchmarking?

So initially I was benchmarking with only 100 files so the performance gain was quite low. Tested with more files and most of the time it is ~2x improvement.

```
func BenchmarkFilesetFiles(b *testing.B) {
	shard := uint32(0)
	numIters := 100
	dir := createDataFilesWithVolumeIndex(nil, dataDirName, testNs1ID, shard, numIters, true, dataFileSuffix, 0)
	defer os.RemoveAll(dir)

	for i := 0; i < b.N; i++ {
		files, err := filesetFiles(filesetFilesSelector{
			fileSetType:    persist.FileSetFlushType,
			contentType:    persist.FileSetDataContentType,
			filePathPrefix: dir,
			namespace:      testNs1ID,
			shard:          shard,
			pattern:        filesetFilePrefix + "*",
		})
		if err != nil {
			b.Fatal(err)
		}

		if len(files) == 0 {
			b.Fatal("no files found")
		}
	}
}
```

`-test.benchtime 3s`
// new
numIters := 1000
BenchmarkFilesetFiles-12    	    2124	   1611317 ns/op
BenchmarkFilesetFiles-12    	    2108	   1619529 ns/op
numIters := 10000
BenchmarkFilesetFiles-12    	     121	  27584583 ns/op
BenchmarkFilesetFiles-12    	     116	  26826172 ns/op
numIters := 20000
BenchmarkFilesetFiles-12    	      21	 146918829 ns/op
BenchmarkFilesetFiles-12    	      18	 188116477 ns/op

//old
numIters := 1000
BenchmarkFilesetFiles-12    	    1189	   3033102 ns/op
BenchmarkFilesetFiles-12    	    1174	   2936289 ns/op
numIters := 10000
BenchmarkFilesetFiles-12    	      55	  56059259 ns/op
BenchmarkFilesetFiles-12    	      58	  54876307 ns/op
numIters := 20000
BenchmarkFilesetFiles-12    	      14	 233025020 ns/op
BenchmarkFilesetFiles-12    	      10	 308521944 ns/op

**Comment by linasm** [comment]
> func BenchmarkFilesetFiles(b *testing.B) {
> 	shard := uint32(0)
> 	numIters := 100
> 	dir := createDataFilesWithVolumeIndex(nil, dataDirName, testNs1ID, shard, numIters, true, dataFileSuffix, 0)
> 	defer os.RemoveAll(dir)

This is benchmarking creation of the files on disk. You need to call `b.ResetTimer()` between benchmark setup and the actual benchmark loop.

**Comment by soundvibe** [comment]
> > func BenchmarkFilesetFiles(b *testing.B) {
> > shard := uint32(0)
> > numIters := 100
> > dir := createDataFilesWithVolumeIndex(nil, dataDirName, testNs1ID, shard, numIters, true, dataFileSuffix, 0)
> > defer os.RemoveAll(dir)
> 
> This is benchmarking creation of the files on disk. You need to call `b.ResetTimer()` between benchmark setup and the actual benchmark loop.

🤦🏻  Thanks for spotting it. Updated results:
```
numIters := 1000
--new
BenchmarkFilesetFiles-12    	    2260	   1600852 ns/op
BenchmarkFilesetFiles-12    	    2282	   1595209 ns/op
--old
BenchmarkFilesetFiles-12    	    1023	   2945657 ns/op
BenchmarkFilesetFiles-12    	    1142	   3255573 ns/op

numIters := 10000
--new
BenchmarkFilesetFiles-12    	     157	  22408915 ns/op
BenchmarkFilesetFiles-12    	     152	  22152958 ns/op
--old
BenchmarkFilesetFiles-12    	      76	  42065111 ns/op
BenchmarkFilesetFiles-12    	      76	  40777160 ns/op

numIters := 20000
--new
BenchmarkFilesetFiles-12    	      62	  51564310 ns/op
BenchmarkFilesetFiles-12    	      69	  48784302 ns/op
--old
BenchmarkFilesetFiles-12    	      30	 103220878 ns/op
BenchmarkFilesetFiles-12    	      30	 104825039 ns/op
```

Still ~2x improvement.

File: src/dbnode/persist/fs/files.go
Comment (linasm) [inline src/dbnode/persist/fs/files.go:1225]: Nit: I'd give more readable names to these vars, eg. `iStart/jStart, iVolume/jVolume`.

File: src/dbnode/persist/fs/files.go
Comment (linasm) [inline src/dbnode/persist/fs/files.go:1240]: I think this `if` is redundant.

File: src/dbnode/persist/fs/files.go
Comment (linasm) [inline src/dbnode/persist/fs/files.go:1233]: Nit: perhaps adding `Sorted` into the name of this function would make it more clear what it does.

**Review by linasm — APPROVED** [review]
LGTM with some comments.

File: src/dbnode/persist/fs/files.go
Comment (soundvibe) [reply src/dbnode/persist/fs/files.go:1240]: I didn't want to change semantics on what is returned from this function. Previously `nil` would be returned if  `filepath.Glob` returns `nil`. Without this `if` check we would be returning empty slice which is not the same thing as before.

Comment (linasm) [reply src/dbnode/persist/fs/files.go:1243]: `findSortedFilesetFiles` is a new function so I guess there can be no existing semantics for it.
And for the `filesetFiles`, there is still an `if` for this purpose:
https://github.com/m3db/m3/pull/3900/files#diff-78d9cf687193bca4cdc4ae73e54059f6a3b3ab4360a627c4bf26c070b8a0f909R1340

Comment (soundvibe) [reply src/dbnode/persist/fs/files.go:1243]: `findSortedFilesetFiles` replaced `findFiles` (which is still present) so its semantics were based on it so that these both functions could be used interchangeably if needed. In Go returning nil slices is quite common and I don't see problems with this approach here.

Comment (linasm) [reply src/dbnode/persist/fs/files.go:1243]: But `findFiles` does not have such check, it returns whatever is returned by `filepath.Glob`...

Comment (soundvibe) [reply src/dbnode/persist/fs/files.go:1243]: Exactly, and `filepath.Glob` returns `nil` if it founds nothing. So `findSortedFilesetFiles`should also return `nil` when it founds nothing.

**Pull request merged**
